nengo_spa/algebras/base.py

```diff
@@ -288,3 +452,82 @@ def zero_element(self, d, sidedness=ElementSidedness.TWO_SIDED):
             Zero element.
         """
         raise NotImplementedError()
+
+
+class AbstractSign(ABC):
+    """Abstract base class for implementing signs for an algebra."""
```

Comment:
The need for this abstraction is a bit unclear to me. It seems like the sign could more simply be represented by a value that can either be -1, 0, or 1 (as in the output of `np.sign`). Then `to_vector` would look something like:
```
if sign == +1:
    return algebra.identity_element(d)
elif sign == -1:
    return algebra.negative_identity_element(d)
elif sign == 0:
    return algebra.zero_element(d)
else:
    raise ValueError("... indeterminant ...")
```
and that would just be a method on the algebra (rather than a separate class). I'm not sure what having this as a class buys us over this alternative? 

A more specific concern I have with this abstraction is it doesn't appear to guarantee that `is_positive` and `is_negative` and `is_zero` are mutually exclusive options.. when I think they should always be?

Reply:
After thinking a bit more I realized this doesn't work for the HRR algebra since `to_vector` needs to handle 3 negative cases in order to get `abs` to work properly. But if we take a step back, `to_vector` is only needed by `abs` and `create_vector`, correct? So couldn't all that casework be the responsibility of those methods? And then you wouldn't really need the sign for anything except to limit the `binding_power` method?

Reply:
* [ ] Yes, I think `HrrAlgebra.create_vector` could just call `abs`. (For the other algebras, this is not sufficient because the sign is usually indefinite.)

* `is_positive`, `is_negative` etc. should indeed be mutually exclusive. Currently, it is up to the implementer to ensure this. I agree, that it might be preferable to ensure this on the API level, but currently I am not sure how to do that while still preserving a nice to use API. I guess, in some way the sign needs to be mapped to an enum. But taking a `sign` and then calling a function to get the "actual" sign value seems not optimal to me. I will think more about this.
* As for why I implemented this abstraction:
** I think it clearly defines one relevant aspect of the vectors. That helps (at least me) to reason about them and to implement other methods such as `abs` and `create_vector` on a higher level of abstraction.
** I also consider this part of the public API, i.e. allowing a user to check the sign of a (maybe randomly generated) vector.
** I also don't really want to include `to_vector` within `create_vector` or `abs` because it would add quite a bit of complexity to those methods while it could actually be nicely factored out (as done right now) with well defined inputs and outputs for `to_vector` and it keeps `create_vector`/`abs` at one level of abstraction.
** The `to_vector` makes also sense on a conceptual level to me because it both the sign can either be represented as one or two numbers or equivalently by the vector. The sign of a number is also a number, analogously the sign of a SPA can be a SPA.